controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,407 @@
+		}
+		existing := pipelineRun.Annotations[FailureCountAnnotationName]
+		if existing == "" {
+			existing = "0"
+		}
+		failureCount, err := strconv.Atoi(existing)
+		if err != nil {
+			log.Error(err, "failed to pass retry count, not retrying")
+			return r.removePipelineFinalizer(ctx, pipelineRun)
+		}
+		failureCount = failureCount + 1
+
+		pipelineRun.Annotations[FailureCountAnnotationName] = strconv.Itoa(failureCount)
+
+		//update the list of successful updates
+		//if any
+		if len(alreadyProcessed) > 0 {
+			pipelineRun.Annotations[AlreadyNudgedAnnotationName] = strings.Join(maps.Keys(alreadyProcessed), ",")
+		}
+		r.EventRecorder.Event(updatedComponent, corev1.EventTypeWarning, ComponentNudgeFailedEventType, fmt.Sprintf("component update failed as a result of a build for %s, retry %d/%d", updatedComponent.Name, failureCount, MaxAttempts))
+
+		if failureCount >= MaxAttempts {
+			// We are at the failure limit, nothing much we can do
+			log.Info("not retrying as max failure limit has been reached", l.Audit, "true")
+			return r.removePipelineFinalizer(ctx, pipelineRun)
+		}
+		log.Info(fmt.Sprintf("failed to update component dependencies, retry %d/%d", failureCount, MaxAttempts))
+		err = r.Client.Update(ctx, pipelineRun)
+		if err != nil {
+			// If we fail to update just return and let requeue handle it
+			// We can't really do anything else
+			// This does mean components may get nudged a second time, but it is idempotent anyway
+			return ctrl.Result{}, err
+		}
+		if immediateRetry {
+			return reconcile.Result{RequeueAfter: time.Millisecond}, nil
+		} else {
+			return reconcile.Result{RequeueAfter: retryTime * time.Duration(failureCount)}, nil
+		}
+	}
+
+	_, err = r.removePipelineFinalizer(ctx, pipelineRun)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Now we need to look for 'stale' pipelines
+	// These are defined as pipelines that are younger than this one, that target the same component
+	// If there are two pushes very close together variances in pipeline run times may mean that the
+	// Older one finishes last. In this case we want to mark the older one as already nudged and remove
+	// The finalizer
+
+	pipelines := tektonapi.PipelineRunList{}
+	err = r.Client.List(ctx, &pipelines, client.InNamespace(pipelineRun.Namespace), client.MatchingLabels{PipelineRunTypeLabelName: PipelineRunBuildType, PipelineRunComponentLabelName: updatedComponent.Name})
+	if err != nil {
+		// I don't think we want to retry this, it should be really rare anyway
+		// And would require an even more complex label based state machine
+		log.Error(err, "failed to check for stale pipeline runs, this operation will not be retried", l.Audit, "true")
+		return ctrl.Result{}, nil
+	}
+	for i := range pipelines.Items {
+		possiblyStalePr := pipelines.Items[i]
+		if possiblyStalePr.Annotations == nil || possiblyStalePr.Annotations[PacEventTypeAnnotationName] != PacEventPushType || possiblyStalePr.Name == pipelineRun.Name {
```

Comment:
Should we also skip pipelines with `NudgeProcessedAnnotationName` annotation?

Reply:
bump

Skipping already-processed pipeline makes sense to me as well